src/axom/quest/detail/MarchingCubesImpl.hpp

```diff
@@ -0,0 +1,763 @@
+      {
+        copyFirstSurfaceCellId(i);
+      }
+    }
+#endif
+
+    m_contourCellCount = m_crossings.empty()
+      ? 0
+      : m_crossings.back().firstSurfaceCellId +
+        num_contour_cells(m_crossings.back().caseNum);
+  }
+
+  void computeContpur() override
+  {
+    auto crossingsView = m_crossings.view();
+
+    /*
+      Reserve contour mesh data space so we can add data without
+      reallocation.
+    */
+    const axom::IndexType contourNodeCount = DIM * m_contourCellCount;
+    m_contourNodeCoords.resize(contourNodeCount);
+    m_contourCellCorners.resize(m_contourCellCount);
+    m_contourCellParents.resize(m_contourCellCount);
+
+    axom::for_all<ExecSpace>(
+      0,
+      m_crossingCount,
+      AXOM_LAMBDA(axom::IndexType iCrossing) {
+        const auto& crossingInfo = crossingsView[iCrossing];
+        const IndexType crossingCellCount =
+          num_contour_cells(crossingInfo.caseNum);
+        SLIC_ASSERT(crossingCellCount > 0);
+
+        // Parent cell data for interpolating new node coordinates.
+        Point cornerCoords[CELL_CORNER_COUNT];
+        double cornerValues[CELL_CORNER_COUNT];
+        get_corner_coords_and_values(crossingInfo.parentCellNum,
+                                     cornerCoords,
+                                     cornerValues);
+
+        /*
+          Create the new cell and its DIM nodes.  New node are on
+          parent cell edges where the edge intersects the isocontour.
+          linear_interp for the exact coordinates.
+
+          TODO: The varying crossingCellCount value may inhibit device
+          performance.  Try grouping m_crossings items that have the
+          same values for crossingCellCount.
+        */
+        for(int iCell = 0; iCell < crossingCellCount; ++iCell)
+        {
+          IndexType contourCellId = crossingInfo.firstSurfaceCellId + iCell;
+          m_contourCellParents[contourCellId] = crossingInfo.parentCellNum;
+          for(int d = 0; d < DIM; ++d)
+          {
+            IndexType contourNodeId = contourCellId * DIM + d;
+            m_contourCellCorners[contourCellId][d] = contourNodeId;
+
+            const int edge = cases_table(crossingInfo.caseNum, iCell * DIM + d);
+            linear_interp(edge,
+                          cornerCoords,
+                          cornerValues,
+                          m_contourNodeCoords[contourNodeId]);
+          }
+        }
+      });
+  }
+
+  // These 4 functions provides access to the look-up table
+  // whether on host or device.  Is there a more elegant way
+  // to put static 1D and 2D arrays on both host and device?  BTNG.
+  template <int TDIM = DIM>
+  AXOM_HOST_DEVICE typename std::enable_if<TDIM == 2, int>::type num_contour_cells(
+    int iCase) const
+  {
+#define _MC_LOOKUP_NUM_SEGMENTS
+#include "marching_cubes_lookup.hpp"
+#undef _MC_LOOKUP_NUM_SEGMENTS
+    SLIC_ASSERT(iCase >= 0 && iCase < 16);
+    return num_segments[iCase];
+  }
```

Comment:
Is there a performance advantage to implementing it this way, w/ internal includes rather than by just including `marching_cubes_lookup.hpp` at the top of this file?

Reply:
Yes.  This is the only way I, with Brian Han's help, could think of to put 2D C arrays on both host and devices without inlining the arrays (which would add hundreds of lines).

Reply:
Could you add a comment explaining this?

Reply:
Should this file use the .inl file extension instead? Would that be more clear?